### PR TITLE
Fix repeated value

### PR DESCRIPTION
**Pull request by tremby**
# Description

Two commits. One adds a failing test case for a bug where in certain conditions the largest value in the srcset list is repeated. The other commit fixes the bug. Full details are in the commit messages.

## Checklist: Bug Fix

- [x] Each commit follows the [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) format: e.g. `chore(readme): fixed typo`. See the end of this file for more information.
- [x] All existing unit tests are still passing (if applicable)
- [x] Add new passing unit tests to cover the code introduced by your PR
- [ ] Update the readme **(n/a)**
- [ ] Update or add any necessary API documentation **(n/a)**
- [x] Add some [steps](#steps-to-test) so we can test your bug fix

## Steps to Test

See the new test for an example failing case.

`buildSrcSet('image.jpg', {}, {widthTolerance: 0.0999, minWidth: 1000, maxWidth: 1200})` was outputting [1000, 1200, 1200].

### CODE REVIEW

**Comment by sherwinski** [comment]
Thanks for the PR @tremby! Overall I think this is a good fix and should be merged. But it also made me wonder how often something like `widthTolerance: 0.0999` actually occurs in the wild. If it's not that sensible of a use case, I'm thinking it would just be better to restrict `widthTolerance` to two decimal places.

**Comment by tremby** [comment]
The 0.0999 isn't really the point; the point is that a bug is possible and this is a fix for it. The bug could also rear its head if there are simply a lot of steps -- this module is multiplying repeatedly by a floating-point number after all.

**Comment by sherwinski** [comment]
> The 0.0999 isn't really the point; the point is that a bug exists

Agreed there, but I was more so trying to understand if you had discovered this out of an actual use case. To be clear, I'm not saying this PR should be changed but I am thinking out loud about if having more guardrails could help prevent potential issues for users.

**Comment by tremby** [comment]
I discovered that "class" of bug when punching in what in my opinion are
some very realistic values for min and max width to show the before/after
example for my other PR. In the new code, right away I found that the
response had a repeated value. That made me scrutinize the code to fix it,
and also made me wonder whether the bug existed in the old code. The bug
didn't exist for the exact same set of inputs on the old code but I very
quickly did find a failing case. Not a realistic one, to be fair, but I
think proof of *a* failing case is enough for a fix, in lieu of proof that
it can't possibly happen in any real-world realistic use case.

That second breaking case is what I added to the tests here in this PR, and
then I added the failing case I originally found as part of the other PR,
and added it to the same test.

I think it's very possible this could happen in real life, especially given
how small the default tolerance is, and therefore how many times it gets
multiplied. If you want me to, I'll write some code to brute-force find a
realistic breaking case, but I don't know why you'd require that!

**Comment by sherwinski** [comment]
>  If you want me to, I'll write some code to brute-force find a realistic breaking case, but I don't know why you'd require that!

Nope, definitely don't want that! 🙂 
Thanks for offering your insights. Someone from the team will go through and formally review the changes within the next day or so.

**Review by ericdeansanchez — COMMENTED** [review]
Hey @tremby 👋 thanks for opening this PR!

File: src/imgix-core-js.js [hunk 1]
```diff
@@ -238,14 +243,12 @@
         return 2 * Math.round(n / 2);
       };
 
-      var prev = minWidth;
-      while (prev < maxWidth) {
-        resolutions.push(ensureEven(prev));
-        prev *= 1 + (INCREMENT_PERCENTAGE * 2);
+      var exact = minWidth;
```
[the remark this answers is not in the document]
Comment (tremby) [reply src/imgix-core-js.js:246]: `nextWidth` isn't really what it means, though. This variable is the non-rounded non-evened running value; the exact width currently under consideration. It's used both in the context of what to round for the "next"/"current" width to add to the array and then in the next loop as the "previous" one's more-exact value for scaling up for the next candidate. IMO `exact` is closer than anything mentioning "next" or "previous". Would `exactWidth` be better?

For my own money, this variable is used for so few lines of code (3, all very close together) that it doesn't need a particularly descriptive name -- one needs to read only a little code to understand its purpose. I think "exact" is very descriptive as it says it's not rounded. Maybe `nonRounded` to make that extra extra clear?

File: src/imgix-core-js.js [hunk 2]
```diff
@@ -224,10 +224,15 @@
 
     // returns an array of width values used during scrset generation
     ImgixClient.prototype._generateTargetWidths = function(widthTolerance, minWidth, maxWidth) {
-      var resolutions = [];
+      var resolutions = [minWidth];
```
[the remark this answers is not in the document]
Comment (tremby) [reply src/imgix-core-js.js:227]: Oops, I hadn't spotted that line. Yes, I'll move it.

File: src/imgix-core-js.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (tremby) [reply src/imgix-core-js.js:247]: Or TBH go the other direction and call it `w` or something! To show it's something temporary, just for the purpose of this small code loop.

File: test/test-buildSrcSet.js [hunk 3]
```diff
@@ -953,6 +953,20 @@ describe('SrcSet Builder:', function describeSuite() {
                     }, Error);
                 });
             });
+
+            describe('with widthTolerance, minWidth, and maxWidth set to promote an edge case', function describeSuite() {
```
[the remark this answers is not in the document]
Comment (tremby) [reply test/test-buildSrcSet.js:957]: But 0.0999 isn't some magic number, nor are 0.09 and 0.10. There will be more failing cases at other values of `widthTolerance`, or for `0.09` or `0.10` for other values of `minWidth` and `maxWidth`. The test isn't really about those specific numbers; it's about coaxing out an edge case which reveals a bug.

[the remark this answers is not in the document]
Comment (tremby) [reply test/test-buildSrcSet.js:957]: This may just be a difference in how we like to describe tests. I like to describe the bug I'm trying to highlight, if possible, whereas perhaps you like to describe particular error-prone circumstances.

I'll change the name of it to what you suggested if you confirm.

File: src/imgix-core-js.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (frederickfogerty) [reply src/imgix-core-js.js:247]: I agree with @tremby here, I'm in favour of something that shows its temporary nature. What about `tempWidth`?

File: test/test-buildSrcSet.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (frederickfogerty) [reply test/test-buildSrcSet.js:957]: This is a hard test case to find a good name for. Overall I'm in agreement with @tremby here. I don't think we should hard-code a value in the test case name, because it doesn't capture the purpose of the test. I think we could improve a bit (as already mentioned) what an "edge case" is, e.g. something like "with widthTolerance, minWidth and maxWidth values that have caused duplicate values in the past"

File: src/imgix-core-js.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (tremby) [reply src/imgix-core-js.js]: I've renamed it to that.

File: test/test-buildSrcSet.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (tremby) [reply test/test-buildSrcSet.js]: I've changed some names accordingly.

**Comment by tremby** [comment]
If/when this is deemed ready, let me squash the tweaks into the respective original commits before merging, please.

**Comment by frederickfogerty** [comment]
Hey @tremby thank you for updating this so quickly 🙌  From my side it's all good, and from our internal discussions, it _seems_ like we are overall happy with this PR. I'd wait on squashing/rebasing this until @ericdeansanchez replies in here, since he's the one currently overseeing imgix-core-js. He's on holiday today but hopefully should get back to you soon!

**Comment by ericdeansanchez** [comment]
I went ahead and squashed/merged this 👍 @tremby just wanted to say thanks again for this fix! We really appreciate it.
